src/batched/dense/impl/KokkosBatched_SVD_Serial_Internal.hpp

```diff
@@ -0,0 +1,355 @@
+#ifndef __KOKKOSBATCHED_SVD_SERIAL_INTERNAL_HPP__
+#define __KOKKOSBATCHED_SVD_SERIAL_INTERNAL_HPP__
+
+
+/// \author Brian Kelley (bmkelle@sandia.gov)
+
+#include "Kokkos_MathematicalFunctions.hpp"
+#include "KokkosBatched_SetIdentity_Internal.hpp"
+#include "KokkosBatched_Givens_Serial_Internal.hpp"
+#include "KokkosBatched_ApplyGivens_Serial_Internal.hpp"
+#include "KokkosBatched_Householder_Serial_Internal.hpp"
+#include "KokkosBatched_ApplyHouseholder_Serial_Internal.hpp"
+
+//Use this macro to handle raw pointer/stride based 2D indexing in this file (just for readability)
+//Requires that for pointer X, the corresponding row/col strides are named Xs0 and Xs1.
+#define SVDIND(arr, i, j) arr[(i) * arr##s0 + (j) * arr##s1]
+#define SVDSWAP(a, b) {auto tmp = a; a = b; b = tmp;}
+
+namespace KokkosBatched {
+
+  ///
+  /// Serial Internal Impl
+  /// ==================== 
+
+  struct SerialSVDInternal {
+
+    //Find the two eigenvalues of [a11 a21 ; a21 a22] by solving the characteristic quadratic.
+    //Since matrix is symmetric these will be real.
+    //NOTE: this is essentially the Wilkinson shift routine already in Batched,
+    //however this is simpler because it exploits the symmetric structure, and the realness of the eigenvalues.
+    template<typename value_type>
+    KOKKOS_INLINE_FUNCTION static void symEigen2x2(value_type a11, value_type a21, value_type a22, value_type& e1, value_type& e2)
+    {
+      value_type a = Kokkos::ArithTraits<value_type>::one();
+      value_type b = -a11 - a22;
+      value_type c = a11 * a22 - a21 * a21;
+      value_type sqrtDet = Kokkos::Experimental::sqrt(b * b - 4 * a * c);
```

Comment:
I know the matrix is supposed to be symmetric but is it worth checking in case someone uses this routine and gets weird behavior out of it?

Reply:
@lucbv The matrix is symmetric by definition, it's always the last 2x2 diagonal block of B^T * B where B is bidiagonal. symEigen2x2 only has a21 as a parameter btw (not a12, it assumes a21 == a12).